### PR TITLE
make coreboot 4.11 platforms buildable under recent OSea (CircleCI builds on debian-11 here)

### PR DESCRIPTION
**Pull request by tlaurion**
~This is WIP.~

In an attempt to investigate why car.data is not build under coreboot 4.11 builds under debian-11, I've decided to check what is going on in Heads' build system, which at the end of the day is supposed to do the right thing until we have something better to use.

To have reproducible builds, we are supposed to have fixed versions of some tools, prior of building anything in the hope of producing reproducible output:
- known make version being same as defined in module definition
- known gawk version being same as defined in module definition
- musl-cross-make toolchain as defined per musl-cross module definition to build all modules **_but coreboot_**
- coreboot version specific build toolchain so coreboot parts are reproducible.
- At that point, we should have reproducible coreboot, reproducible kernel, reproducible initrd containing all Heads related built modules (tools.cpio, heads.cpio being bundled) which should be hopefully all being reproducible again.
- Have a reproducible ROM for a specific commit ID, if and only the precedent are all true.

If initrd is not reproducible, then check individual modules output under hashes.txt. Check logs for configure output of faulty module. Diffscope. Patch. Redo.

I first noticed that local gawk was only built if make was also in wrong local version, and then noticed that most of the make calls to build all modules were actually not using the locally built make version but the host deployed version, from both configure logs and build logs under build/logs/*.

This is an attempt to correct the situation. 

My make fu not being optimal, please review and comment, and/or propose another PR with proper fixes. Also.... Please educate me if i'm doing stupid things when compared to original Makefile (which was obviously also doing stupid things.... as you can see by my attempt of improving it).

The DATE fixation Makefile statement was embedded under make version checking under Makefile as well. This is also fixed per current PR.

Still TODO: We lost rom hash output under hashes.txt. Not sure since when and why, yet. Things got broken when gawk local build was added, or around that time.

### CODE REVIEW

**Comment by tlaurion** [comment]
It was coincidental that it worked for debian-10, since locally deployed make was the good one per OS deployed versions (make 4.2.1):

```
root@89a0503e542f:~# make --version
GNU Make 4.2.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

```

Also to be noted that under debian-10, gawk was not installed by default, the Makfile test depending on something else then gawk if not found...

On debian-10, we see that gawk is reported to be a "bad version", but was still not compiled: https://circleci.com/api/v1.1/project/github/tlaurion/heads/3735/output/103/0?file=true&allocation-id=61fbece2b47315079ce0ff17-0-build%2F30054AE1

Also to be noted that the locally built make and gawk were not properly being propagated for make calls for all other modules.
Basically, Heads was still using host deployed versions.

**Comment by tlaurion** [comment]
Rebased on master to only include changes related to support debian-11

**Comment by tlaurion** [comment]
Now, builds fail on 

```
/root/project/build/make-4.2.1/make -C "/root/project/build/coreboot-4.11/kgpe-d16_workstation" -C /root/project/build/coreboot-4.11 obj=/root/project/build/coreboot-4.11/kgpe-d16_workstation CROSS=/root/project/crossgcc/bin/i386-linux-musl- IASL=/root/project/build/coreboot-4.11/util/crossgcc/xgcc/bin/iasl DOTCONFIG=/root/project/build/coreboot-4.11/kgpe-d16_workstation/.config BUILD_TIMELESS=1 CFLAGS_x86_32=-fdebug-prefix-map=/root/project=heads -gno-record-gcc-switches -Wno-error=packed-not-aligned CFLAGS_x86_64=-fdebug-prefix-map=/root/project=heads -gno-record-gcc-switches -Wno-error=packed-not-aligned -j36 --max-load 16
make[1]: Entering directory '/root/project/build/coreboot-4.11'
/root/project/build/make-4.2.1/make oldconfig
KCONFIG_STRICT= kgpe-d16_workstation/util/kconfig/conf --oldconfig src/Kconfig
#
# configuration written to /root/project/build/coreboot-4.11/kgpe-d16_workstation/.config
#
kgpe-d16_workstation/util/kconfig/conf --silentoldconfig src/Kconfig
    HOSTCC     util/romcc/romcc (this may take a while)
gcc -g  -Wall -o kgpe-d16_workstation/util/romcc/romcc /root/project/build/coreboot-4.11/util/romcc/romcc.c
    GEN        build.h
mv kgpe-d16_workstation/build.ht kgpe-d16_workstation/build.h
    CC         bootblock/arch/x86/walkcbfs.o
/root/project/build/coreboot-4.11/util/crossgcc/xgcc/bin/i386-elf-gcc -MMD -Isrc -Isrc/include -Isrc/commonlib/include -Ikgpe-d16_workstation -I3rdparty/vboot/firmware/include -include src/include/kconfig.h -include src/include/rules.h -include src/commonlib/include/commonlib/compiler.h -I3rdparty -D__BUILD_DIR__=\"kgpe-d16_workstation\" -Isrc/superio/aspeed -Isrc/arch/x86/include -D__ARCH_x86_32__ -pipe -g -nostdinc -std=gnu11 -nostdlib -Wall -Wundef -Wstrict-prototypes -Wmissing-prototypes -Wwrite-strings -Wredundant-decls -Wno-trigraphs -Wimplicit-fallthrough -Wstrict-aliasing -Wshadow -Wdate-time -Wtype-limits -Wvla -fno-common -ffreestanding -fno-builtin -fomit-frame-pointer -ffunction-sections -fdata-sections -fno-pie -fno-delete-null-pointer-checks -Wno-packed-not-aligned -fconserve-stack -Wnull-dereference -Wreturn-type -Wno-unused-but-set-variable -Werror -Os -I3rdparty/vboot/firmware/2lib/include -I3rdparty/vboot/firmware/lib21/include -fdebug-prefix-map=/root/project=heads -gno-record-gcc-switches -Wno-error=packed-not-aligned -MT kgpe-d16_workstation/bootblock/arch/x86/walkcbfs.o -D__PRE_RAM__ -D__BOOTBLOCK__ -c -o kgpe-d16_workstation/bootblock/arch/x86/walkcbfs.o src/arch/x86/walkcbfs.S
    GEN        generated/assembly.inc
printf '#include "src/cpu/amd/car/cache_as_ram.inc"\n' > kgpe-d16_workstation/generated/assembly.inc
/bin/sh: 1: cannot create kgpe-d16_workstation/generated/assembly.inc: Directory nonexistent
make[1]: *** [src/arch/x86/Makefile.inc:61: kgpe-d16_workstation/generated/assembly.inc] Error 2
make[1]: Leaving directory '/root/project/build/coreboot-4.11'
make: *** [Makefile:503: /root/project/build/coreboot-4.11/kgpe-d16_workstation/.build] Error 1

CircleCI received exit code 0
```

I am redoing build on top of debian-10 with the modified Makefile to see if it breaks here: https://app.circleci.com/pipelines/github/tlaurion/heads/962/workflows/8ed9bcc0-4539-4bed-88e2-6296ee203692

**Comment by tlaurion** [comment]
builds on top of debian-10 with same modified Makefile worked here: https://app.circleci.com/pipelines/github/tlaurion/heads/962/workflows/7af97058-b209-4e3d-8a4f-2250cb1fd25c

So now i'm even more puzzled.
If we are using local built make and gawk, building musl-cross-make to build modules inside of coreboot, and build coreboot with its own coreboot-version based toolchain..... Then why:

- Debian-10 based kgpe-d16_workstation coreboot-11 [fails](https://app.circleci.com/pipelines/github/tlaurion/heads/961/workflows/51bc0444-e29d-42f8-804e-ee3dabd0ea8c/jobs/3902) here
  - [success build log ](https://circleci.com/api/v1.1/project/github/tlaurion/heads/3937/output/103/0?file=true&allocation-id=61fd8103191925212603350f-0-build%2F7793805D)
  - (logs can be downloaded from artifacts tab)
- Debian-11 based kgpe-d16_workstation coreboot-10 [succeeds](https://app.circleci.com/pipelines/github/tlaurion/heads/962/workflows/7af97058-b209-4e3d-8a4f-2250cb1fd25c/jobs/3937) here
  - [failure build log](https://circleci.com/api/v1.1/project/github/tlaurion/heads/3902/output/103/0?file=true&allocation-id=61fd6b7823ef4a1910321bae-0-build%2F7DDBCDF1)
  - (logs can be downloaded from artifacts tab)

**Comment by tlaurion** [comment]
The next run fixes it magically. My eyes do not see what needs to be fixed under coreboot Makefile, Makefile.inc to fix that race condition...

@mrchromebox? Same applies in current master for build.h

So the parts generating build.h and assembly.inc are fragile in coreboot 4.11. What fixed it in later coreboot versions?

**Comment by tlaurion** [comment]
Seems like https://review.coreboot.org/c/coreboot/+/42883 might fix this

**Comment by tlaurion** [comment]
And https://review.coreboot.org/c/coreboot/+/11904

Will test...

**Comment by tlaurion** [comment]
Todo: debian-10 reports bad gawk version even if good. Fix it

**Comment by tlaurion** [comment]
Good news. Builds are now stable on debian-11 for coreboot 4.11 boards since https://github.com/osresearch/heads/pull/1106/commits/bd0d080820e7bc7adb0988cf50ed943f6770e84f and do not need reruns when there is concurrency problems (at least for 36 cores as under CircleCI) since local gawk and make are now actually properly propagated down the Heads buildsystem.

Notice at beginning of builds on debian-11+ :
```
2022-02-05 14:26:43-05:00 Wrong OS deployed make detected: 4.3. Building and using 4.2.1...
2022-02-05 14:26:43-05:00 Wrong OS deployed gawk detected: 5.1.0. Building and using 4.2.1...
2022-02-05 14:26:43-05:00 Local built HEADS_GAWK only if different then provided by OS: /root/project/build/gawk-4.2.1/gawk
2022-02-05 14:26:43-05:00 Local built HEADS_MAKE only if different then provided by OS: /root/project/build/make-4.2.1/make
2022-02-05 14:26:43-05:00 Heads build system will call make from now on as: MAKE: /root/project/build/make-4.2.1/make.
```

@MrChromebox a clean build, just to make sure, is happening at https://app.circleci.com/pipelines/github/tlaurion/heads/965/workflows/51ea86e3-1ad1-4514-8998-c1888b8fc2bd

Should be ready to merge. But if you could flash librem_l1um rom from CircleCI, as usual, that would be amazing.

**Comment by tlaurion** [comment]
@mrchromebox : https://4107-103208611-gh.circle-artifacts.com/0/build/librem_l1um/heads-librem_l1um-v5.0.1-98-g510ce2b.rom

**Comment by MrChromebox** [comment]
> @MrChromebox : https://4107-103208611-gh.circle-artifacts.com/0/build/librem_l1um/heads-librem_l1um-v5.0.1-98-g510ce2b.rom

boot tested successfully on L1UM
